### PR TITLE
sql: re-execute distributed query as local for some errors

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit teaches the main query code path (i.e. ignoring sub- and
post-queries) to retry distributed plans as local in some cases. In
particular, we use this retry mechanism if:
- the error is SQL retryable (i.e. it'll have a high chance of success
during the local execution)
- no data has been pushed to the result writer by the distributed query
(this shouldn't be a frequent scenario since most SQL retryable errors
are likely to occur during the plan setup / before any data can be
produced by the query).

This retry mechanism allows us to hide transient network problems,
and - more importantly - in the multi-tenant model it allows us to go
around the problem when "not ready" SQL instance is being used for
DistSQL planning (e.g. the instance might have been brought down, but
the cache on top of the system table hasn't been updated accordingly).
I believe that no matter the improvements that we can make to the
instance cache, there will also be a window (which should hopefully
getting smaller - according to David T it's currently 45s but he hopes
to bring it down to 7s or so) during which the instance cache is stale,
so DistSQL planner could use "not ready" instances.

The rationale for why it is ok to do this retry is that we create
brand-new processors that aren't affiliated to the distributed plan
that was just cleaned up. It's worth mentioning that the planNode
tree couldn't haven been reused in this way, but if we needed to
execute any planNodes directly, then we would have to run such a plan
in a local fashion. In other words, the fact that we had a
distributed plan initially guarantees that we don't have any
planNodes to be concerned about.

Possible downside to this approach is that it increases overall query
latency, so ideally we wouldn't plan on "not ready" instances in the
first place (and we have issues to improve there), but given that we now
fully parallelize the setup of distributed plans, the latency increase
should be bound, assuming that most retryable errors occur during the
distributed plan setup.

Addresses: #100578.
Epic: None

Release note: None

### CODE REVIEW

**Review by msirek — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105451)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @yuzefovich)

___
*[`pkg/sql/distsql_physical_planner.go` line 4590 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-NYsncVh6b71Y4qT8DFx:-NYsncVh6b71Y4qT8DFy:b43u1mh) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc4233cb041295080e46af8986e38a292e00d1/pkg/sql/distsql_physical_planner.go#L4590)):*
> ```Go
> ) *PlanningCtx {
> 	distribute := distributionType == DistributionTypeAlways || (distributionType == DistributionTypeSystemTenantOnly && evalCtx.Codec.ForSystemTenant())
> 	planCtx := dsp.newPlanningCtxForLocal(evalCtx, planner, localityFiler)
> ```

Why does this always build a local planning context, even when `distribute` is `true`?
___
*[`pkg/sql/distsql_running.go` line 1238 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-NYsrkED1N0ICJcMxrkV:-NYsrkED1N0ICJcMxrkW:b-knjou6) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc4233cb041295080e46af8986e38a292e00d1/pkg/sql/distsql_running.go#L1238)):*
> ```Go
> 	r.status = execinfra.NeedMoreRows
> 	r.closed = false
> 	r.stats = stats
> ```

Please reset `r.progressAtomic`. Maybe something like: 
`atomic.StoreUint64(r.progressAtomic, 0)`

Also, if `r.discardRows` is `true`, we could be retrying with a non-initialized egress counter, so maybe add something like this:
```go
		if r.egressCounter != nil {
			r.egressCounter = NewTenantNetworkEgressCounter()
		}
```
It appears `discardRows` is only used for testing, but we'd still want the counter to be correct for tests.
___
*[`pkg/sql/distsql_running.go` line 1963 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-NYsjT350nje6R8pIzUv:-NYsjT350nje6R8pIzUw:bgymqoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc4233cb041295080e46af8986e38a292e00d1/pkg/sql/distsql_running.go#L1963)):*
> ```Go
> 		// brand-new processors that aren't affiliated to the distributed plan
> 		// that was just cleaned up. It's worth mentioning that the planNode
> 		// tree couldn't haven been reused in this way, but if we needed to
> ```

nit: "couldn't haven" --> "couldn't have"
___
*[`pkg/sql/distsql_running.go` line 1996 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-NYslKLeAazidGvPZzBb:-NYslKLeAazidGvPZzBc:b-ysockv) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc4233cb041295080e46af8986e38a292e00d1/pkg/sql/distsql_running.go#L1996)):*
> ```Go
> 			return
> 		}
> 		finalizePlanWithRowCount(ctx, localPlanCtx, localPhysPlan, localPlanCtx.planner.curPlan.mainRowCount)
> ```

Would it be useful to check again for context cancellation here before calling `dsp.Run`, in case it happened while generating the new physical plan?
Also, please add the line:
`recv.expectedRowsRead = int64(localPhysPlan.TotalEstimatedScannedRows)`

**Review by yuzefovich — COMMENTED** [review]
Thanks for taking a look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-:-NYu6Kis3gHHEaSBIWtR:bubxju8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @msirek)

___
*[`pkg/sql/distsql_running.go` line 1238 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-NYsrkED1N0ICJcMxrkV:-NYu4ziVATTAIdi08Gwx:b-omsoh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc4233cb041295080e46af8986e38a292e00d1/pkg/sql/distsql_running.go#L1238)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Please reset `r.progressAtomic`. Maybe something like: 
`atomic.StoreUint64(r.progressAtomic, 0)`

Also, if `r.discardRows` is `true`, we could be retrying with a non-initialized egress counter, so maybe add something like this:
```go
		if r.egressCounter != nil {
			r.egressCounter = NewTenantNetworkEgressCounter()
		}
```
It appears `discardRows` is only used for testing, but we'd still want the counter to be correct for tests.

</blockquote></details>

Nice catch, done. For `progressAtomic` I somewhat deliberately was being lazy (this progress tracking is broken in the vectorized engine due #55758), but it's better to be thorough here.
___
*[`pkg/sql/distsql_running.go` line 1996 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-NYslKLeAazidGvPZzBb:-NYu5a-s6tnLtlAKEw2_:bm35i03) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc4233cb041295080e46af8986e38a292e00d1/pkg/sql/distsql_running.go#L1996)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Would it be useful to check again for context cancellation here before calling `dsp.Run`, in case it happened while generating the new physical plan?
Also, please add the line:
`recv.expectedRowsRead = int64(localPhysPlan.TotalEstimatedScannedRows)`
</blockquote></details>

Updated `expectedRowsRead` (I think it shouldn't matter since this field remains unchanged from the distributed run and I'd guess the local estimate would be the same, but in any case it's better to be explicit).

Checking context cancellation doesn't seem that useful - we don't do that in the distributed run, and this local execution will detect it sooner if the context has just been canceled during the plan generation.
___
*[`pkg/sql/distsql_physical_planner.go` line 4590 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-NYsncVh6b71Y4qT8DFx:-NYu3g_BB-Ad5hcgGXBq:b-36ed3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc4233cb041295080e46af8986e38a292e00d1/pkg/sql/distsql_physical_planner.go#L4590)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Why does this always build a local planning context, even when `distribute` is `true`?
</blockquote></details>

I thought it would be more clear to have this helper method that sets up the planning context for local execution, and then the caller can extend it for distributed execution if necessary. It seems like it is confusing, so I removed the new helper in favor of using the existing constructor method.

**Review by msirek — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-:-NYuN-TtCxwfsHHDBN7a:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach)

**Comment by cucaroach** [comment]
Will this fix https://github.com/cockroachdb/cockroach/issues/102839 I wonder?

**Comment by yuzefovich** [comment]
I think it'll fix `select-job` failures (where the internal executor uses DistSQL to run a query and it fails), but [this failure](https://github.com/cockroachdb/cockroach/issues/102839#issuecomment-1540489672) is just wrong (perhaps it was another manifestation of already fixed by #101477 although seems unlikely).

Tommy, I'll wait for your review here too since this change deserves extra scrutiny.

File: pkg/sql/distsql_running.go [hunk 1]
```diff
@@ -1486,6 +1504,7 @@ func (r *DistSQLReceiver) Push(
 	if commErr := r.resultWriterMu.row.AddRow(r.ctx, r.row); commErr != nil {
 		r.handleCommErr(commErr)
 	}
+	r.dataPushed = true
```
Comment (cucaroach) [inline pkg/sql/distsql_running.go:1507]: What about the pushMeta early return path above?

File: pkg/sql/distsql_running_test.go [hunk 2]
```diff
@@ -972,3 +973,114 @@ CREATE TABLE child (
 		sqlDB.Exec(t, fmt.Sprintf(`%[1]sINSERT INTO child VALUES (%[2]d, %[2]d, %[2]d)`, prefix, id))
 	}
 }
+
+// TestDistributedQueryErrorIsRetriedLocally verifies that if a query with a
+// distributed plan results in a SQL retryable error, then it is rerun as local
+// transparently.
+func TestDistributedQueryErrorIsRetriedLocally(t *testing.T) {
```
Comment (cucaroach) [inline pkg/sql/distsql_running_test.go:980]: Can we add a test that makes sure the cluster setting works to disable this?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105451)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @msirek)

___
*[`pkg/sql/distsql_running.go` line 1507 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105451#gh-1244371336:-NYynQwTEfU8GyeI6Q3R:b-jqkewk) ([raw file](https://github.com/cockroachdb/cockroach/blob/336ec0f8194a75ac2aeac27249957a27905d3f91/pkg/sql/distsql_running.go#L1507)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

What about the pushMeta early return path above?
</blockquote></details>

I audited the code, and I _think_ it'd be ok to ignore all metadata types for the purposes of the retry since it doesn't make it to the result writer, the only exception is this `MetadataResultWriter`, and users of that don't call `PlanAndRun`, so they aren't getting this automatic retry. That said, it seems like a good idea to consider any piece of non-error metadata as "data pushed" as long as it makes to the result writer, so I adjusted `dataPushed` boolean in that case. We need to exclude the error metadata because we know it'll be communicated when we want this retry to kick in, and we'll override it with the result of the local execution.
___
*[`pkg/sql/distsql_running_test.go` line 980 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105451#gh-1244381759:-NYyoBqa7mqz1LjlKKlZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/336ec0f8194a75ac2aeac27249957a27905d3f91/pkg/sql/distsql_running_test.go#L980)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Can we add a test that makes sure the cluster setting works to disable this?
</blockquote></details>

Done.

File: pkg/sql/distsql_running.go [hunk 3]
```diff
@@ -1911,6 +1953,61 @@ func (dsp *DistSQLPlanner) PlanAndRun(
...
+			// the connection is toast. We also give up if the context
+			// cancellation has already occurred.
+			return
+		}
+		if !pgerror.IsSQLRetryableError(distributedErr) && !flowinfra.IsFlowRetryableError(distributedErr) {
+			// Only re-run the query if we think there is a high chance of a
+			// successful local execution.
+			return
+		}
+		log.VEventf(ctx, 1, "encountered an error when running the distributed plan, re-running it as local: %v", distributedErr)
```
Comment (cucaroach) [inline pkg/sql/distsql_running.go:1993]: Is this enough observability to see how frequently this is happening or should we have some telemetry too?

**Review by cucaroach — APPROVED** [review]

**Review by yuzefovich — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105451#-:-NZ20hgh9W_WdeyzpXK4:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @msirek)

___
*[`pkg/sql/distsql_running.go` line 1993 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105451#gh-1244844304:-NZ20XJg9jLuMJ36g7z3:b-qbir8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e22342c84bf32ba9f06e1034ad7e0b904f7963e/pkg/sql/distsql_running.go#L1993)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Is this enough observability to see how frequently this is happening or should we have some telemetry too?
</blockquote></details>

Hm, not sure how we're going to use it, but why not, added a couple of counters.

**Comment by cucaroach** [comment]
Should we backport this to 23.1?

**Comment by yuzefovich** [comment]
Seems a bit risky, what's the rationale?

**Comment by cucaroach** [comment]
> Seems a bit risky, what's the rationale?

I was suffering under the delusion that 23.1 had distsql turned on for internal executor and this was the cause of some 23.1 flakes but I've recovered ;-)

**Pull request merged**